Improve support for homematic garage covers

guillempages opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Currently the HmIP garage cover in the homematic integration only provide a switch for turning the light on/off and a sensor to check the status, but no cover entity to be able to actually control the entity.
This PR adds a cover entity that both shows the status and can be used to open/close the cover.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
homematic:
  interfaces:
    wireless:
      host: homematic
      port: 2001
      resolvenames: false
    ip:
      host: homematic
      port: 2010
    thermostatgroups:
      host: homematic
      port: 9292
      path: "/groups"
  hosts:
    ccu2:
      host: homematic
      username: !secret homematic_user
      password: !secret homematic_password
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #35270 
- This PR is related to issue: 
- Link to documentation pull request: 
- This PR requires PR https://github.com/danielperna84/pyhomematic/pull/308 in pyhomematic to be merged (and pyhomematic version to be bumped)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

guillempages committed:
Improve support for homematic garage covers

Add a cover entity for HmIP-MOD-HO and HmIP-MOD-TM devices

pvizeli commented on homeassistant/components/homematic/cover.py:
```diff
@@ -48,11 +54,20 @@ def set_cover_position(self, **kwargs):
-        if self.current_cover_position is not None:
+        """Return whether the cover is closed."""
+        if isinstance(self._hmdevice, IPGarage):
+            return self._hmdevice.is_closed()
```
You can not do IO inside the property. You need to read it out form data cache

guillempages replied:
Thanks for the tip. Now I'm even getting the updates (which was the missing piece in the code before)

pvizeli commented on homeassistant/components/homematic/cover.py:
```diff
@@ -36,6 +39,9 @@ def current_cover_position(self):
 
         None is unknown, 0 is closed, 100 is fully open.
         """
+        if isinstance(self._hmdevice, IPGarage):
```
We use everwhere class names

pvizeli replied:
Let use the same schema, `HM_GARAGE = ("IPGarage", )` and check it again this

guillempages committed:
Implement PR comments

Use class name instead of isinstance to detect IPGarage devices
Use the cached values instead of explicitly reading the sensor.

guillempages commented:
@pvizeli Thanks for the first review. I've done the changes you proposed; how does it look like?

Any chance of getting this into 0.110?

balloob commented on homeassistant/components/homematic/cover.py:
```diff
@@ -48,11 +54,20 @@ def set_cover_position(self, **kwargs):
-        """Return if the cover is closed."""
+        """Return whether the cover is closed."""
+        if self._hmdevice.__class__.__name__ in HM_GARAGE:
+            return self._hmdevice.is_closed(self._hm_get_state())
```
Does this do I/O ? You are not allowed to do I/O in properties.

pvizeli replied:
Yes, like I said in my prev comments, you need use the values from data cache

guillempages replied:
This call is not doing any I/O. It is just making the values from the state to a Boolean. It behaves like a static method, but did not want to make it really static in order to avoid including the pyhomematic class here.

I could just do the mapping here (IIRC 0 is closed, 1 and 2 is in-between and 3 is open; I don't remember exactly);.I just thought that the mapping belonged in the library and not here.

balloob commented on homeassistant/components/homematic/cover.py:
```diff
@@ -36,6 +39,9 @@ def current_cover_position(self):
 
         None is unknown, 0 is closed, 100 is fully open.
         """
+        if self._hmdevice.__class__.__name__ in HM_GARAGE:
```
It feels weird to hack this in like this. 

Instead you should create a new class `HMGarage` and inside `setup_platform` instantiate the right class

pvizeli replied:
Yeah, 2x entity class like i.e. Nuki

guillempages replied:
Ok, I misunderstood the original comment by @pvizeli. Will do that, but I can't do anything before next week

guillempages replied:
I've now created the HMGarage class, and I'm using this for the methods that need to be different. Could you take another look?

guillempages committed:
Create subclass for garage covers

guillempages commented:
@pvizeli @balloob I've implemented the requested changes; could you take a look again and see if it is OK as it is implemented now?

balloob commented on homeassistant/components/homematic/cover.py:
```diff
@@ -20,7 +23,10 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     devices = []
     for conf in discovery_info[ATTR_DISCOVER_DEVICES]:
-        new_device = HMCover(conf)
+        if conf[ATTR_DEVICE_TYPE] in HM_GARAGE:
```
Why does the user need to configure this, can't we auto detect it based on available data ?

guillempages replied:
It is auto configured; it's added automatically do the discovery data when the integration is set up. I had to pass the information from the generic homematic initialization, where all devices are detected, to the individual entity types setup. Is there a better way? As far as I could see, there is other information being added there.
